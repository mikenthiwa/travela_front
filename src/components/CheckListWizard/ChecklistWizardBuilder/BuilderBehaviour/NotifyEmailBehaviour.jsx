import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as yup from 'yup';
import Modal from '../../../modal/Modal';
import Helper from '../BuilderOptions/helper';

const schema = yup.object({
  recipient: yup.string().email('Recipient must be a valid email').required('Recipient is required'),
  template: yup.string().min(25, 'Template must be more than 25 characters').required('Template is required'),
});

class NotifyEmailBehaviour extends Component {
    
  state = {
    recipient: '',
    template: '',
    errors: null,
    touched: [],
    shouldOpen: false,
  }

  componentDidMount () {
    this.resetState();
    this.validate();
  }

  
  toggleModal = () => {
    const { shouldOpen } = this.state;
    this.setState({ shouldOpen: !shouldOpen });
  };

  
  handleChange = ({ target: { value, name } }) => {
    this.setState({ [name]: value }, this.validate);
  }
  
  handleCreateTemplate = () => {
    const { recipient, template } = this.state;
    const { behaviour, handleBehaviour } = this.props;
    
    handleBehaviour({ ...behaviour, payload: { ...behaviour.payload, recipient, template } });
    
    this.toggleModal();
  }

  handleTouched = (e) => {
    const { touched } = this.state;
    this.setState({ touched: [...touched, e.target.name] });
  }
  
  isErrored = (name) => {
    const { errors, touched } = this.state;
    return touched.includes(name) && errors && errors[name];
  }
  
  resetState() {
    const { behaviour: { payload: { recipient, template } } } = this.props;
    this.setState({
      recipient: recipient || '',
      template: template || '',
      touched: [],
      shouldOpen: false,
    });
  }

  validate() {
    const { recipient, template } = this.state;
    try {
      schema.validateSync({ recipient, template }, { abortEarly: false });
      this.setState({ errors: null });
    } catch (error) {
      const errors = error.inner.reduce((prev, curr) => ({ ...prev, [curr.path]: curr.message }), {});
      this.setState({ errors });
    }
  }

  renderTemplateModal = () => {
    const { recipient, template, errors, shouldOpen } = this.state;
    return (
      <Modal
        customModalStyles="notification-email"
        closeModal={this.toggleModal} width="500px"
        visibility={shouldOpen ? 'visible' : 'invisible'}
        title={`${recipient ? 'Edit' : 'Create'} Template`}
      >
        <div className="email-template-modal-wrapper">
          <div className="template-inputs-wrapper">
            <input
              type="email"
              id="emailToSend"
              name="recipient"
              className="email-template-input recipient"
              placeholder="ex. example@andela.com"
              value={recipient}
              onChange={this.handleChange}
              onKeyDown={Helper.disableInputUndoActions}
              onBlur={this.handleTouched}
            />
            {this.isErrored('recipient') && <span className="input-error">{errors['recipient']}</span>}
          </div>
          <div className="template-inputs-wrapper">
            <textarea
              name="template"
              className="email-template-input template"
              placeholder="Your message here"
              value={template}
              onChange={this.handleChange}
              onKeyDown={Helper.disableInputUndoActions}
              onBlur={this.handleTouched}
            />
            {this.isErrored('template') && <span className="input-error">{errors['template']}</span>}
          </div>
          <div className="email-template-button-wrapper">
            <button
              className="email-template-button"
              type="button"
              disabled={!!errors}
              onClick={this.handleCreateTemplate}
            >
            Create
            </button>
          </div>
        </div>
      </Modal>
    );
  }

  render() {
    const { recipient } = this.state;
    return (
      <div className="email-template-behaviour">
        {this.renderTemplateModal()}
        <button
          type="button"
          onClick={this.toggleModal}
          className="create-email-template-button"
        >
          {`${recipient ? 'Edit' : 'Create'} Template`}
        </button>
      </div>
    );
  }
}

NotifyEmailBehaviour.propTypes = {
  behaviour: PropTypes.object.isRequired,
  handleBehaviour: PropTypes.func.isRequired,
};

export default NotifyEmailBehaviour;
