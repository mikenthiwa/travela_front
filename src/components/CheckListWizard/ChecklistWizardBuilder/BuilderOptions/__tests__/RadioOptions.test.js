import React from 'react';
import { shallow } from 'enzyme';
import RadioOption from '../RadioOptions';
import NotifyEmailBehaviour from '../../BuilderBehaviour/NotifyEmailBehaviour';

jest.mock('react-redux', () => ({
  connect: (state, dispatch) => {
    return component => component;
  }
}));

const props = {
  item: {
    id: 'kldskd',
    name: 'are you serious?',
    behaviour: {
      type: 'NOTIFY_EMAIL',
      payload: 'john@example.com'
    }
  },
  updateBehaviour: jest.fn(),
  deleteQuestion: jest.fn(),
};

const props3 = {
  behaviour: { type: 'NOTIFY_EMAIL', payload: {} },
  handleBehaviour: jest.fn(),
  closeModal: jest.fn(), 
  openModal: jest.fn(),
  shouldOpen: false,
};

const mockNotify = props3 => {
  return shallow(<NotifyEmailBehaviour {...props3} />);
};

const setup = props => {
  return shallow(<RadioOption {...props} />);
};

const setup2 = props => {
  return mount(<RadioOption {...props} />);
};
describe('<RadioOption />', () => {
  let wrapper;
  let secondWrapper;
  let wrapper3;
  beforeEach(() => {
    wrapper = setup(props);
    secondWrapper = setup2(props);
    wrapper3 = mockNotify(props3);
  });
  it('should render RenderCheckbox component properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should render RenderCheckbox component properly', () => {
    expect(secondWrapper).toMatchSnapshot();
  });

  it('should handle update behaviour', () => {
    const mockEvents = { target: { value: 'Passport?'} };
    const btn = secondWrapper.find('#option-name-input');
    btn.simulate('change', mockEvents);
    wrapper.instance().changeBehaviour(wrapper.behaviour);
    expect(wrapper.updateBehaviour).toBeCalled;
  });

  it('should handle delete icon onChange', () => {
    const mockEvents = { target: { value: 'Passport?'} };
    const deleteIcon = secondWrapper.find('button#option-del-icon');
    deleteIcon.simulate('click', mockEvents);
    expect(wrapper.deleteQuestion).toBeCalled;
  });

  it('should handle show behaviour btn', () => {
    const mockEvents = jest.fn();
    const btn = secondWrapper.find('.set-behaviour-btn');
    btn.simulate('click', mockEvents);
    expect(wrapper.showListOfBehaviours).toBeCalled;
  });

  it('should handle show behaviour btn', () => {
    
    const wrapper = mount(<RadioOption {...props} />);
    wrapper.find('.set-behaviour-btn').first().simulate('click');
    const input = wrapper3.find('.create-email-template-button');
    input.simulate('click');
    expect(input).toHaveLength(1);
    expect(wrapper.instance().props.updateBehaviour).toBeCalled;
  });
});
