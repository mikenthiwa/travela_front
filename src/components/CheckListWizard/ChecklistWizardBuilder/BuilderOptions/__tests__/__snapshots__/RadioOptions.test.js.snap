// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioOption /> should render RenderCheckbox component properly 1`] = `
<div>
  <div>
    <p>
      Enter Option undefined
    </p>
    <div
      className="prompt-item"
    >
      <input
        className="prompt-input"
        id="option-name-input"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Yes"
        type="text"
        value="are you serious?"
      />
      <DeleteIcon
        id="option-del-icon"
        onClick={[Function]}
      />
    </div>
    <div
      className="behaviour-container"
    >
      <div
        className="set-behaviour-btn-container"
      >
        <button
          className="set-behaviour-btn"
          onClick={[Function]}
          type="button"
        >
          Set behaviour
        </button>
      </div>
      <div>
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`<RadioOption /> should render RenderCheckbox component properly 2`] = `
<div>
  <div>
    <p>
      Enter Option undefined
    </p>
    <div
      className="prompt-item"
    >
      <input
        className="prompt-input"
        id="option-name-input"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Yes"
        type="text"
        value="are you serious?"
      />
      <button
        className="builder-delete-icon"
        id="option-del-icon"
        onClick={[Function]}
        type="button"
      >
        <img
          alt="delete-icon"
          src="delete-icon.svg"
        />
      </button>
    </div>
    <div
      className="behaviour-container"
    >
      <div
        className="set-behaviour-btn-container"
      >
        <button
          className="set-behaviour-btn"
          onClick={[Function]}
          type="button"
        >
          Set behaviour
        </button>
      </div>
      <div>
        <div />
      </div>
    </div>
  </div>
</div>
`;
